### PR TITLE
Issue243 standalone

### PR DESCRIPTION
**Pull request by humphd**

### CODE REVIEW

**Comment by humphd** [comment]
You can test this by loading `src/hosted.html` in your browser, and then you can do this in the console:

```js
var bramble = Bramble.createInstance();
bramble.increaseFontSize();
bramble.showSidebar();
bramble.useMobilePreview();
...
```

File: src/bramble-api.js [hunk 1]
```diff
@@ -0,0 +1,325 @@
...
+        // Whether to transfer ownership of ArrayBuffers or not
+        var allowArrayBufferTransfer;
+
+        if (typeof div === "object"  && !(div instanceof HTMLElement)) {
+            options = div;
+            div = null;
+        }
+        options = options || {};
+
+        var id = self._id = "bramble-" + UUID.generate();
```
Comment (sedge) [inline src/bramble-api.js]: This seems a little over-engineered to me. We won't have more than one instance on a page, so couldn't this be hardcoded without the UUID?

File: src/bramble-api.js [hunk 2]
```diff
@@ -0,0 +1,325 @@
...
+        var provider = options.provider || new Filer.FileSystem.providers.Memory();
+        var fs = self.fs = new Filer.FileSystem({provider: provider});
+
+        function startEvents(win) {
+            window.addEventListener("message", function(e) {
+                var data = parseEventData(e.data);
+
+                // When Bramble asks for initial content, reply but don't bother providing any
+                if (data.type === "bramble:init") {
+                    win.postMessage(JSON.stringify({type: "bramble:init", source: null}), "*");
```
Comment (sedge) [inline src/bramble-api.js]: I guess this was a quick fix?

**Comment by humphd** [comment]
@sedge rebasing deleted your comments, my responses:
* re: using `'bramble-' + UUID`, it just makes sure we have a unique element name, and since anyone can put this in any page, it doesn't hurt.  NOTE: I have to include a UUID lib for other stuff, so it was free--I wouldn't have done this just for that :)
* re: passing the initial source to Bramble--yes, we'll change all this.  I suspect we'll just do it using the `fs` instance directly.  We need to ponder this.  For now, we continue to delegate to the Bramble instance.

File: Gruntfile.js [hunk 3]
```diff
@@ -216,6 +216,23 @@ module.exports = function (grunt) {
                     exclude: ["text!config.json"],
                     uglify2: {} // https://github.com/mishoo/UglifyJS2
                 }
+            },
+            iframe: {
+                // Standalone dist/bramble.js iframe api
+                options: {
+                    name: 'bramble/almond',
```
Comment (sedge) [inline Gruntfile.js]: `almond`?

File: src/extensions/default/bramble/lib/RemoteCommandHandler.js [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+
+    // Built-in Brackets Commands
+    function _bracketsCommand(command) {
+        // Make sure the last-focused editor gets focus before executing
+        EditorManager.getActiveEditor().focus();
+        CommandManager.execute(Commands[command]);
+    }
+
+    // Custom Bramble commands
+    function _brambleCommand(command) {
```
Comment (sedge) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: This is a very elegant refactor. Much needed -.-

File: src/extensions/default/bramble/lib/RemoteCommandHandler.js [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            return;
+        }
+
+        switch(remoteRequest.commandCategory) {
+        case "brackets":
+            _bracketsCommand(remoteRequest.command);
+            break;
+// todo - what is this?
+//        case "editorCommand":
+//  Editor[msgObj.command](msgObj.params);
```
Comment (sedge) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: I think this was Keni's way of proxying brackets commands directly to the editor.

Comment (sedge) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: With a little more digging, it looks like he was using it to set the indentation spacing:

https://github.com/mozilla/thimble.webmaker.org/blob/bramble/public/friendlycode/js/fc/ui/bramble-proxy.js#L112

https://github.com/mozilla/thimble.webmaker.org/blob/bramble/public/friendlycode/js/fc/ui/bramble-proxy.js#L214-L217

File: src/extensions/default/bramble/main.js [hunk 6]
```diff
@@ -326,7 +265,7 @@ define(function (require, exports, module) {
                     fn: data.fn,
                     value: typeof value !== "object" ? value : undefined
                 }), "*");
-            });
+            }, false);
```
Comment (sedge) [inline src/extensions/default/bramble/main.js]: I feel like you might have explained this to me before. Doesn't this parameter default to `false` if it isn't specified?

File: src/extensions/default/bramble/main.js [hunk 7]
```diff
@@ -352,16 +291,15 @@ define(function (require, exports, module) {
...
+            window.removeEventListener("message", _getInitialDocument, false);
+            window.addEventListener("message", RemoteCommandHandler.handleRequest, false);
 
             var fileHTML = FileSystem.getFileForPath("/index.html");
             var fileCSS  = FileSystem.getFileForPath("/style.css");
             var fileJS   = FileSystem.getFileForPath("/script.js");
 
             // Write the HTML file and block on it being done.
-            fileHTML.write(data.source ? data.source : defaultHTML,
+            fileHTML.write(data.source ? data.source : defaultHTML, {blind: true},
```
Comment (sedge) [inline src/extensions/default/bramble/main.js]: I struggled to find docs for this option. Can I assume it forces the write?

File: src/hosted.html [hunk 8]
```diff
@@ -10,22 +10,36 @@
...
+        body {
             margin: 0;
-            padding: 0;
         }
 
-        iframe: {
-            width: 100%;
-            height: 100%;
+        .full {
+            height: 100vh;
```
Comment (sedge) [inline src/hosted.html]: I've never seen this unit before. Neat!

**Comment by sedge** [comment]
Code looks good! I'm getting jshint errors on `build-browser` though:

screen shot 2015-06-08 at 10 05 21 pm

**Comment by sedge** [comment]
Lastly, I only tested on chrome. `undo`, `redo` and `enableJavascript` didn't work. The `enableJavascript` might have been user error though... see if you can repeat it.

File: src/extensions/default/bramble/lib/RemoteCommandHandler.js (hunk 5 above)
Comment (humphd) [inline src/extensions/default/bramble/lib/RemoteCommandHandler.js]: I *think* we can just dump this, do you agree?  I feel safer letting Brackets increase/decrease the font size the way Luke has it in his UI mockup vs. what we had before with small/medium/large.

File: src/extensions/default/bramble/main.js (hunk 6 above)
Comment (humphd) [inline src/extensions/default/bramble/main.js]: It's an old habit, see https://developer.mozilla.org/en-US/docs/Web/API/EventTarget/addEventListener and the third arg.

File: src/extensions/default/bramble/main.js (hunk 7 above)
Comment (humphd) [inline src/extensions/default/bramble/main.js]: Yeah, without it you run into issues where it tries to compare against an in memory cache.  See https://github.com/adobe/brackets/blob/master/src/file/FileUtils.js#L96-L108.  It just forces our write to work, despite what's on the filesystem.

File: src/hosted.html (hunk 8 above)
Comment (humphd) [inline src/hosted.html]: It's magic!

File: Gruntfile.js (hunk 3 above)
Comment (humphd) [inline Gruntfile.js]: Almond is a slimmed down shim for requirejs that you can use in deployment when you just want to use requirejs modules to make a standalone thing.  It's mostly about space-saving in production files.

**Comment by humphd** [comment]
The `undo` and `redo` bits are somehow related to focus...I don't know if it's possible to test them from the console, because you necessarily have focus in the console vs. CodeMirror editor when you press enter.  I think in an app they will work, but we need to test.  I'm not sure what `enableJavascript` not working was all about, I'll test tomorrow.

**Comment by humphd** [comment]
Closing in favour of https://github.com/humphd/brackets/pull/349

**Pull request closed without merging**